olive/passes/onnx/mnb_to_qdq.py

```diff
@@ -0,0 +1,307 @@
+import math
+from pathlib import Path
+from typing import TYPE_CHECKING, Any, Dict
+
+import numpy as np
+import onnx
+
+from olive.hardware.accelerator import AcceleratorSpec
+from olive.model import ONNXModelHandler
+from olive.model.utils import resolve_onnx_path
+from olive.passes import Pass
+from olive.passes.onnx.common import get_external_data_config, model_proto_to_olive_model
+from olive.passes.onnx.onnx_dag import OnnxDAG
+from olive.passes.pass_config import PassConfigParam
+
+if TYPE_CHECKING:
+    from numpy.typing import NDArray
+
+logger = logging.getLogger(__name__)
+
+
+class MatMulNBitsToQDQ(Pass):
+    @classmethod
+    def _default_config(cls, accelerator_spec: AcceleratorSpec) -> Dict[str, PassConfigParam]:
+        return {
+            "use_transpose_op": PassConfigParam(
+                type_=bool,
+                # TODO(jambayk): decide whether to enable this by default or not
+                # False gives same output on arm Mac/Windows, but not on x64 Linux/Windows
+                default_value=True,
+                description=(
+                    "Whether to use a Transpose operator after the DequantizeLinear operator. If False, the weight"
+                    " initializer will be transposed instead."
+                ),
+            ),
+            **get_external_data_config(),
+        }
+
+    def _run_for_config(
+        self, model: ONNXModelHandler, config: Dict[str, Any], output_model_path: str
+    ) -> ONNXModelHandler:
+        output_model_path = resolve_onnx_path(output_model_path, Path(model.model_path).name)
+
+        # create a dag from the model
+        dag = OnnxDAG.from_model_path(model.model_path)
+        # remove unnecessary identity nodes
+        dag.remove_identity_nodes()
+
+        num_modified = 0
+        for node_name in dag.get_node_names():
+            op_type = dag.get_node_op_type(node_name)
+            if op_type != "MatMulNBits":
+                continue
+
+            node_inputs = dag.get_node_inputs(node_name)
+            # only deal with the constant matmul case for now
+            if not all(dag.is_initializer(i_name) and not dag.is_input(i_name) for i_name in node_inputs[1:]):
+                continue
+
+            graph_idx = dag.get_graph_idx(node_name)
+
+            # original output proto
+            node_output = dag.get_node_outputs(node_name)[0]
+            is_model_output = dag.is_output(node_output)
+            node_output_proto = None
+            if dag.get_io(node_output).proto:
+                node_output_proto = dag.get_io(node_output).proto[-1]
+            node_attributes = dag.get_node_attributes(node_name)
+            K = node_attributes["K"]  # noqa: N806
+            N = node_attributes["N"]  # noqa: N806
+            block_size = node_attributes["block_size"]
+            num_k_blocks = math.ceil(K / block_size)
+
+            # only deal with 4 bits for now
+            if node_attributes["bits"] != 4:
+                continue
```

Comment:
Add a info message indicating that this is only dealing with 4 bits.

Reply:
INT4 only or FP4/NF4 are fine?

Reply:
int4 only. Matmulnbits only does int4 4bit quantization.

Reply:
done. Added log and updated comment.